openpathsampling/pathmover.py

```diff
@@ -1244,6 +1286,146 @@ def move(self, globalstate):
+        the PathMovers to choose from
+    weights : list of floats
+        the relative weight of each PathMover (does not need to be normalized)
+    """
+
+    def _selector(self, globalstate):
+        if self.weights is None:
+            weights = [1.0] * len(self.movers)
+        else:
+            weights = self.weights
+
+        # this is implemented by setting all weights locally to zero that
+        # correspond to movers that will potentially fail since the required
+        # input ensembles are not present in the globalstate
+
+        present_ensembles = globalstate.ensembles
+
+        for idx, mover in enumerate(self.movers):
+            for ens in mover.input_ensembles:
+                if ens not in present_ensembles:
+                    # ens might be required but is not present
+                    weights[idx] = 0.0
+
+        return weights
+
+class FirstAllowedMover(RandomSelectionMover):
+    """
+    Chooses a first mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in enumerate(self.movers):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
+
+class LastAllowedMover(RandomSelectionMover):
+    """
+    Chooses the last mover that has samples in all required ensembles.
+
+    A mover can only safely be run, if all inputs can be satisfied. This will pick
+    the first mover from the list where all ensembles from input_ensembles are
+    found.
+
+    Attributes
+    ----------
+    movers : list of PathMover
+        the PathMovers to choose from
+    """
+
+    def _selector(self, globalstate):
+        weights = [1.0] * len(self.movers)
+
+        present_ensembles = globalstate.ensembles
+
+        found = False
+
+        for idx, mover in reversed(enumerate(self.movers)):
+            if not found:
+                for ens in mover.input_ensembles:
+                    if ens not in present_ensembles:
+                        # ens might be required but is not present
+                        weights[idx] = 0.0
+
+                if weights[idx] > 0.0:
+                    found = True
+            else:
+                weights[idx] = 0.0
+
+        return weights
```

Comment:
Okay, I see what you're doing, but I think we should eventually find a better solution to this. This is still a pretty nonintuitive way to approach the problem, and feels a bit kludgy. Honestly, the better solution is probably just to keep state in the simulator, and rather than doing these contortions to avoid keeping state. There's nothing equilibrium about the bootstrapping sampling, so these hacks to make it act like it is are just leaving us with a bunch of path movers that are *really* bad ideas for most situations. (These, and the partial acceptance sequential, and I think we've done a few other silly things over time.) It is cruft that will confuse contributors.